File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/ticket/fragment/EventsFragment.java
package de.tum.in.tumcampusapp.component.ui.ticket.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.io.Serializable;
import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.other.generic.adapter.EqualSpacingItemDecoration;
import de.tum.in.tumcampusapp.component.ui.ticket.EventsController;
import de.tum.in.tumcampusapp.component.ui.ticket.adapter.EventsAdapter;
import de.tum.in.tumcampusapp.component.ui.ticket.model.Event;
import de.tum.in.tumcampusapp.component.ui.ticket.model.EventType;
import de.tum.in.tumcampusapp.component.ui.ticket.model.Ticket;
import de.tum.in.tumcampusapp.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class EventsFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener {

    private static final String KEY_EVENT_TYPE = "type";

    private RecyclerView mRecyclerView;
    private SwipeRefreshLayout mSwipeLayout;
    private TextView mPlaceholderTextView;

    private EventType mEventType;
    private EventsController mEventsController;

    private Callback<List<Event>> mEventsCallback = new Callback<List<Event>>() {
        @Override
        public void onResponse(@NonNull Call<List<Event>> call,
                               @NonNull Response<List<Event>> response) {
            List<Event> events = response.body();
            if (events == null) {
                return;
            }

            mEventsController.addEvents(events);
            //loadEventsFromDatabase(); // TODO
            if (mEventType == EventType.ALL) {
                showEvents(events);
            }

            mSwipeLayout.setRefreshing(false);
        }

        @Override
        public void onFailure(@NonNull Call<List<Event>> call, @NonNull Throwable t) {
            Utils.log(t);
            Utils.showToast(getContext(), R.string.error_something_wrong);
            mSwipeLayout.setRefreshing(false);
        }
    };

    private Callback<List<Ticket>> mTicketsCallback = new Callback<List<Ticket>>() {
        @Override
        public void onResponse(@NonNull Call<List<Ticket>> call, Response<List<Ticket>> response) {
            List<Ticket> tickets = response.body();
            if (tickets == null) {
                return;
            }
            mEventsController.addTickets(tickets);
            // TODO
            //loadEventsFromDatabase();
            //showEvents(events);

            if (mEventType == EventType.BOOKED) {
                List<Event> bookedEvents = mEventsController.getBookedEvents();
                showEvents(bookedEvents);
            }

            mSwipeLayout.setRefreshing(false);
        }

        @Override
        public void onFailure(@NonNull Call<List<Ticket>> call, @NonNull Throwable t) {
            Utils.log(t);
            Utils.showToast(getContext(), R.string.error_something_wrong);
            mSwipeLayout.setRefreshing(false);
        }
    };

    public static EventsFragment newInstance(Serializable eventType) {
        EventsFragment fragment = new EventsFragment();
        Bundle args = new Bundle();
        args.putSerializable(KEY_EVENT_TYPE, eventType);
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, 
                             ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.events_fragment, container, false);

        mRecyclerView = view.findViewById(R.id.activity_events_list_view);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        
        int spacing = Math.round(getResources().getDimension(R.dimen.material_card_view_padding));
        mRecyclerView.addItemDecoration(new EqualSpacingItemDecoration(spacing));
        
        mSwipeLayout = view.findViewById(R.id.event_refresh);
        mSwipeLayout.setOnRefreshListener(this);

        mPlaceholderTextView = view.findViewById(R.id.placeholder_text_view);
        
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        // Create the events controller here as the context is only created when the
        // fragment is attached to an activity
        mEventsController = new EventsController(getContext());

        if (getArguments() != null) {
            mEventType = (EventType) getArguments().getSerializable(KEY_EVENT_TYPE);

            if (mEventType != null) {
                List<Event> events = loadEventsFromDatabase(mEventType);
                showEvents(events);
            }
        }
    }

    private List<Event> loadEventsFromDatabase(EventType type) {
        List<Event> events;

        if (type == EventType.ALL) {
            events = mEventsController.getEvents();
        } else {
            events = mEventsController.getBookedEvents();
        }

        return events;
    }

    private void showEvents(List<Event> events)  {
        boolean isEmpty = events.isEmpty();
        mRecyclerView.setVisibility(isEmpty ? View.GONE : View.VISIBLE);
        mPlaceholderTextView.setVisibility(isEmpty ? View.VISIBLE : View.GONE);

        if (!events.isEmpty()) {
            EventsAdapter adapter = new EventsAdapter(getContext(), events);
            mRecyclerView.setAdapter(adapter);
            adapter.notifyDataSetChanged();
        } else {
            mPlaceholderTextView.setText(mEventType.getPlaceholderResId());
        }
    }

    @Override
    public void onRefresh() {
        mEventsController.getEventsAndTicketsFromServer(mEventsCallback, mTicketsCallback);
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/ticket/fragment/EventsFragment.kt
package de.tum.`in`.tumcampusapp.component.ui.ticket.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.other.generic.adapter.EqualSpacingItemDecoration
import de.tum.`in`.tumcampusapp.component.ui.ticket.EventsController
import de.tum.`in`.tumcampusapp.component.ui.ticket.EventsViewModel
import de.tum.`in`.tumcampusapp.component.ui.ticket.adapter.EventsAdapter
import de.tum.`in`.tumcampusapp.component.ui.ticket.model.Event
import de.tum.`in`.tumcampusapp.component.ui.ticket.model.EventType
import de.tum.`in`.tumcampusapp.component.ui.ticket.model.Ticket
import de.tum.`in`.tumcampusapp.utils.Utils
import de.tum.`in`.tumcampusapp.utils.observeNonNull
import kotlinx.android.synthetic.main.fragment_events.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class EventsFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var eventType: EventType
    private lateinit var eventsController: EventsController

    private val eventsCallback = object : Callback<List<Event>> {
        override fun onResponse(call: Call<List<Event>>, response: Response<List<Event>>) {
            val events = response.body() ?: return
            eventsController.storeEvents(events)

            if (this@EventsFragment.isDetached.not()) {
                eventsRefreshLayout.isRefreshing = false
            }
        }

        override fun onFailure(call: Call<List<Event>>, t: Throwable) {
            val c = requireContext()
            Utils.log(t)
            Utils.showToast(c, R.string.error_something_wrong)

            if (this@EventsFragment.isDetached.not()) {
                eventsRefreshLayout.isRefreshing = false
            }
        }
    }

    private val ticketsCallback = object : Callback<List<Ticket>> {
        override fun onResponse(call: Call<List<Ticket>>, response: Response<List<Ticket>>) {
            val tickets = response.body() ?: return
            eventsController.insert(*tickets.toTypedArray())
            eventsRefreshLayout.isRefreshing = false
        }

        override fun onFailure(call: Call<List<Ticket>>, t: Throwable) {
            val c = requireContext()
            Utils.log(t)
            Utils.showToast(c, R.string.error_something_wrong)
            eventsRefreshLayout.isRefreshing = false
        }
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_events, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        with(view) {
            eventsRecyclerView.setHasFixedSize(true)
            eventsRecyclerView.layoutManager = LinearLayoutManager(context)
            eventsRecyclerView.itemAnimator = DefaultItemAnimator()
            eventsRecyclerView.adapter = EventsAdapter(context)

            val spacing = Math.round(resources.getDimension(R.dimen.material_card_view_padding))
            eventsRecyclerView.addItemDecoration(EqualSpacingItemDecoration(spacing))

            eventsRefreshLayout.setOnRefreshListener(this@EventsFragment)
            eventsRefreshLayout.setColorSchemeResources(
                    R.color.color_primary,
                    R.color.tum_A100,
                    R.color.tum_A200
            )
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        eventType = arguments?.getSerializable(KEY_EVENT_TYPE) as EventType
        eventsController = EventsController(context)

        val factory = EventsViewModel.Factory(eventsController, eventType)
        val viewModel = ViewModelProviders.of(this, factory).get(EventsViewModel::class.java)

        viewModel.events.observeNonNull(this) { showEvents(it) }
    }

    private fun showEvents(events: List<Event>) {
        val isEmpty = events.isEmpty()
        eventsRecyclerView.visibility = if (isEmpty) View.GONE else View.VISIBLE
        placeholderTextView.visibility = if (isEmpty) View.VISIBLE else View.GONE

        if (events.isNotEmpty()) {
            val adapter = eventsRecyclerView.adapter as EventsAdapter
            adapter.update(events)
        } else {
            placeholderTextView.setText(eventType.placeholderResId)
        }
    }

    override fun onRefresh() {
        eventsController.getEventsAndTicketsFromServer(eventsCallback, ticketsCallback)
    }

    companion object {

        private const val KEY_EVENT_TYPE = "type"

        @JvmStatic
        fun newInstance(eventType: EventType): EventsFragment {
            return EventsFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(KEY_EVENT_TYPE, eventType)
                }
            }
        }
    }

}
